agent/engine/execcmdagent/manager_linux.go

```diff
@@ -0,0 +1,209 @@
+// +build linux
+
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+package execcmdagent
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/cihub/seelog"
+	"github.com/docker/docker/api/types"
+
+	apicontainer "github.com/aws/amazon-ecs-agent/agent/api/container"
+	apitask "github.com/aws/amazon-ecs-agent/agent/api/task"
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient"
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient/dockerapi"
+	"github.com/aws/amazon-ecs-agent/agent/utils/retry"
+)
+
+// RestartIfStopped restarts the ExecCommandAgent in the container passed as parameter, only for ExecCommandAgent-enabled tasks.
+// The status of the ExecCommandAgent in the container is retrieved using a docker exec inspect call, using the dockerExecID
+// stored in container.GetExecCommandAgentMetadata().DockerExecID.
+//
+// If container.GetExecCommandAgentMetadata() is nil, or the ExecCommandAgent is still running, no action is taken.
+// To actually restart the ExecCommandAgent, this function invokes this instance's Start method.
+func (m *manager) RestartIfStopped(ctx context.Context, client dockerapi.DockerClient, task *apitask.Task, container *apicontainer.Container, containerId string) (bool, error) {
+	if !task.IsExecCommandAgentEnabled() {
```

Comment:
can this and below check be done in the parent method to fail early?

Reply:
I thought about it, and I do it for Start, can do it for this one too, but I still want to keep this checks inside the functions because the engine is not very well encapsulated and in the future someone might call these functions without first checking if exec is enabled.

Reply:
Actually, this is implicitly done because the health check process skips non-exec enabled tasks: https://github.com/aws/amazon-ecs-agent/pull/2648/files#diff-b9aa12c3425c34d6f330f28407f779f6R281

Reply:
we could later think about moving the exec methods in engine to exec agent manager

Reply:
I'm ok with the overhead of these checks -- in case this function gets called from the wrong context we have some simple guard rails.